File: src/main/java/io/jjong/algorithm/word/PalindromeWord.java
package io.jjong.algorithm.word;

/**
 * create on 2022/12/03. create by IntelliJ IDEA.
 *
 * <p> 클래스 설명 </p>
 * <p> {@link } and {@link }관련 클래스 </p>
 *
 * @author Jongsang Han(Henry)
 * @version 1.0
 * @see
 * @since 1.0
 */
public class PalindromeWord {

  public static boolean isPalindrome(String s) {
    // i 는 처음부터, j는 뒤에서 부터
    int i=0, j=s.length()-1;
    while (i < j) {
      while (!Character.isLetterOrDigit(s.charAt(i)) && i < j ) {
        ++i;
        System.out.println("i = "+ i);
      }
      while (!Character.isLetterOrDigit(s.charAt(j)) && i < j ) {
        --j;
        System.out.println("j = "+ j);
      }
      System.out.println("start comparing word,  i : " + i + ", j : " + j);
      if (Character.toLowerCase(s.charAt(i++)) != Character.toLowerCase(s.charAt(j--))) {
        return false;
      }
    }
    return true;
  }

  public static void main(String[] args){
    String word = "!! ablevelba   !!";
    boolean palindrome = isPalindrome(word);
    System.out.println("is palindrome : " + palindrome);
  }

}

File: src/main/kotlin/io/jjong/algorithm/word/PalindromeWord.kt
package io.jjong.algorithm.word

/**
 * 영숫자만 비교(대소문자 무시)했을 때 회문인지 검사한다. 양끝에서 투 포인터로 좁혀간다.
 */
object PalindromeWord {
    fun isPalindrome(s: String): Boolean {
        var i = 0
        var j = s.length - 1
        while (i < j) {
            // 영숫자가 아니면 건너뛴다.
            while (!s[i].isLetterOrDigit() && i < j) {
                ++i
                println("i = $i")
            }
            while (!s[j].isLetterOrDigit() && i < j) {
                --j
                println("j = $j")
            }
            println("start comparing word,  i : $i, j : $j")
            if (s[i].lowercaseChar() != s[j].lowercaseChar()) {
                return false
            }
            ++i
            --j
        }
        return true
    }
}

fun main() {
    val word = "!! ablevelba   !!"
    val palindrome = PalindromeWord.isPalindrome(word)
    println("is palindrome : $palindrome")
}
